proof/refine/ARM/Reply_R.thy

```diff
@@ -1081,27 +1024,841 @@ lemma sc_replies_relation_replyNext_update:
+  supply opt_mapE[rule del]
+  apply (frule (1) sc_replies_relation_sc_with_reply_heap_path)
+  apply (prop_tac "replyNexts_of s' nrp = Some rp")
+   apply (simp add: sym_refs_replyNext_replyPrev_sym[symmetric])
+  apply (drule sc_with_reply_SomeD, clarsimp)
+  apply (prop_tac "the (sc_replies_of s scp) = sc_replies sc")
+  using heap_ls_unique sc_replies_relation_prevs_list' apply blast
+  apply simp
+  apply (frule (2) heap_ls_next_in_list)
+  apply (frule (2) sym_refs_replySCs_of_None[where rp=nrp])
+   apply (erule replyNexts_Some_replySCs_None)
+  apply (drule_tac x=scp in spec)
+  apply (clarsimp simp: vs_heap_simps)
+  apply (frule (2) heap_ls_next_takeWhile_append[where p=rp])
+  apply (frule in_list_decompose_takeWhile[where x=nrp])
+  apply (frule in_list_decompose_takeWhile[where x=rp])
+  apply auto
+  done
+
+lemma sc_replies_middle_reply_sc_None:
+  "\<lbrakk>sym_refs (state_refs_of s); valid_replies s; sc_with_reply rp s = Some scp;
+    (sc_replies_of s |> hd_opt) scp \<noteq> Some rp; sc_at scp s; reply_at rp s\<rbrakk> \<Longrightarrow>
+   reply_sc_reply_at ((=) None) rp s"
+  apply (clarsimp simp: obj_at_def is_sc_obj_def is_reply)
+  apply (rename_tac ko n reply; case_tac ko; clarsimp)
+  apply (rename_tac sc n)
+  apply (drule sc_with_reply_SomeD1)
+  apply (clarsimp simp: vs_heap_simps opt_map_Some)
+  apply (case_tac "sc_replies sc"; simp)
+   apply (clarsimp simp: sc_replies_sc_at_def obj_at_def)
+  apply (rule ccontr)
+  apply (clarsimp simp: reply_sc_reply_at_def obj_at_def)
+  apply (case_tac "reply_sc reply"; simp)
+  apply (drule sym_refs_sc_replies_sc_at)
+   apply (fastforce simp: reply_sc_reply_at_def obj_at_def)
+  apply (rename_tac p)
+  apply clarsimp
+  apply (prop_tac "sc_replies_sc_at ((\<lambda>rs. rp \<in> set rs)) p s")
+   apply (clarsimp simp: sc_replies_sc_at_def obj_at_def)
+   apply (metis list.set_intros(1))
+  apply (drule valid_replies_sc_replies_unique)
+   apply fastforce
+  apply (prop_tac "scp=p")
+   apply blast
+  using sc_replies_sc_atE by fastforce
+
+(* FIXME RT: move to AInvs *)
+lemma sc_with_reply_Some_sc_at:
+  "\<lbrakk>sc_with_reply rp s = Some scp; valid_objs s\<rbrakk> \<Longrightarrow> sc_at scp s"
+  by (fastforce dest!: sc_with_reply_SomeD elim: valid_sched_context_size_objsI
+                 simp: obj_at_def is_sc_obj)
+
+(* FIXME RT: from Invariants_AI; rephrase replies *)
+lemma valid_objs_sc_replies_reply_at:
+  "\<lbrakk>valid_objs s; sc_replies_sc_at (\<lambda>rs. rp \<in> set rs) sc_ptr s\<rbrakk> \<Longrightarrow> reply_at rp s"
+  by (fastforce simp: sc_replies_sc_at_def obj_at_def valid_sched_context_def list_all_def
+                dest: valid_sched_context_objsI)
+
+lemma sc_with_reply_Some_reply_at:
+  "\<lbrakk>sc_with_reply rp s = Some scp; valid_objs s\<rbrakk> \<Longrightarrow> reply_at rp s"
+  by (fastforce dest: sc_with_reply_SomeD1 valid_objs_sc_replies_reply_at)
+(* end FIXME RT: move to AInvs *)
```

Comment:
does this comment apply to the previous or next lemma?

Reply:
This is the closing comment for FIXME move three lemmas above.